doc: update "enable-authentication-and-restriction.md" doc, and examples of `jwt-auth` and `wolf-rbac`. (#1018)
apache/apisix-ingress-controller: docs/en/latest/practices/enable-authentication-and-restriction.md

## Attributes

> ### Authentication methods

Reviewer: For me, I think we don't need the `methods` explanation 😄

> #### `keyAuth`

Reviewer: How about using `Key Auth`, `JWT Auth` as title style?

Reviewer 2: Agree.

> Consumers add their key either in a header `apikey` to authenticate their requests. For more information about `keyAuth`, please refer to [APISIX jwt-auth](https://apisix.apache.org/docs/apisix/plugins/key-auth/).

Reviewer: I notice you use `either` here, then where we could use except for `header`?

> <details>

Reviewer: 😂 May I know what's `details` usage here?

Author: ![3KYX3F0OV7J9)E59TCVYK{5](https://user-images.githubusercontent.com/79972061/168740386-394eafdd-088e-4e76-af6a-9427513cc3fe.png)
![O)KPZ{9BSQK2V 4@}6E8 HG](https://user-images.githubusercontent.com/79972061/168740400-0854532a-3722-44e2-9f4a-04011ded90be.png)

Reviewer 3: Do we have specific reason to use this? If so, could you add it to the style guide on when to use this?

Author: My idea is that there are many authentication methods we haven't written in. If we add configuration to each one later, it may appear very bloated. I think the display effect will be better if we use it. Users can choose the configuration content they care about. What do you think?

  <summary>keyAuth yaml configure</summary>

```yaml
apiVersion: apisix.apache.org/v2beta3
kind: ApisixConsumer
metadata:

>   name: ${name}

Reviewer: I would suggest we keep using the same style when passing variables, cc @navendu-pottekkat @hf400159

Reviewer 4: yes, I've added this to the next edition of the writing guide last week, and I'll be releasing the new style guide this Friday.

Reviewer 3: @AlinsRan Does the `${name}` here refer to a value the user has to set? Or is it somehow a variable they are passing?

In the past we have been using `name: user-configured-name` or some placeholder like that.

// @juzhiyuan @hf400159

Author: This is a variable that users can customize. Do I need to follow the style of the past?

spec:
  authParameter:
    keyAuth:
      value:
        key: ${key} #required
```

</details>

> #### `basicAuth`

Reviewer 2: ```suggestion
#### Basic Auth
```

> Consumers add their key either in a header `Authentication` to authenticate their requests.For more information about `basicAuth`, please refer to [APISIX basic-auth](https://apisix.apache.org/docs/apisix/plugins/basic-auth/).

Reviewer: `either`

<details>
  <summary>basicAuth yaml configure</summary>

```yaml
apiVersion: apisix.apache.org/v2beta3
kind: ApisixConsumer
metadata:
  name: ${name}
spec:
  authParameter:
    basicAuth:
      value:
        username: ${username} #required
        password: ${password} #required
```

> </details>

Reviewer 2: I think after each example, we need to give users a tip about using the `valueRef` field to reference a K8s Secret object so that we can avoid the hardcoded sensitive data in the ApisixConsumer object.

Author: Solved.

> #### `jwtAuth`

Reviewer 2: ```suggestion
#### JWT Auth

```

### Authentication

#### Key Auth

> Consumers add their key either in a header or query string parameter to authenticate their requests. For more information about `Key Auth`, please refer to [APISIX key-auth](https://apisix.apache.org/docs/apisix/plugins/key-auth/).  

Reviewer 5: ```suggestion
Consumers add their key either in a header or query string parameter to authenticate their requests. For more information about `Key Auth`, please refer to [APISIX key-auth plugin](https://apisix.apache.org/docs/apisix/plugins/key-auth/).  

```

        secret: "${secret}"                              #optional
        public_key: "${public_key}"                      #optional, required when algorithm attribute selects RS256 algorithm.
        private_key: "{private_key}"                     #optional, required when algorithm attribute selects RS256 algorithm.
        algorithm: "${HS256 | HS512 | RS256}"            #optional
        exp: ${ 86400 | token's expire time, in seconds} #optional
        algorithm: ${true | false}                       #optional
```

</details>

#### `Wolf RBAC`

To use wolfRbac authentication, you need to start and install [wolf-server](https://github.com/iGeeky/wolf/blob/master/quick-start-with-docker/README.md). For more information about `Wolf RBAC`, please refer to [APISIX wolf-rbac plugin](https://apisix.apache.org/zh/docs/apisix/plugins/wolf-rbac/).  
Also, we can using the `secretRef` field to reference a K8s Secret object so that we can avoid the hardcoded sensitive data in the ApisixConsumer object. For reference Secret use example, please refer to the [key-auth-reference-secret-object](#key-auth-reference-secret-object).

> :::note This plugin will add several API

Reviewer: ```suggestion
:::note This plugin will add several APIs

```

Reviewer 5: @AlinsRan PTAL

### How to enable `Authentication`

#### Enable `keyAuth`

The following is an example. The `keyAuth` is enabled on the specified route to restrict user access.  

> * Creates a ApisixConsumer, and set the attributes of plugin `key-auth`:

Reviewer: ```suggestion
* Creates an ApisixConsumer, and set the attributes of plugin `key-auth`:
```

EOF
```

> * Creates a ApisixRoute, and enable plugin `key-auth`:

Reviewer: ```suggestion
* Creates an ApisixRoute, and enables plugin `key-auth`:

```

EOF
```

* Requests from foo:

```shell
kubectl  exec -it -n ${namespace of Apache APISIX} ${pod of Apache APISIX}  -- curl http://127.0.0.1:9080/anything -H 'Host: httpbin.org' -H 'apikey:foo-key' -i
```

```shell
HTTP/1.1 200 OK
...
```

##### Key Auth reference Secret object

<details>
  <summary>ApisixRoute with keyAuth consumer using secret example</summary>

* Creates a `Secret` object:

```shell
kubectl apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: foovalue
data:
  key: Zm9vLWtleQ==
EOF
```

> * Creates a ApisixConsumer and reference `Secret` object:

Reviewer: ```suggestion
* Creates an ApisixConsumer and reference `Secret` object:

```

```shell
kubectl apply -f - <<EOF
apiVersion: apisix.apache.org/v2beta3
kind: ApisixConsumer
metadata:
  name: foo
spec:
  authParameter:
    keyAuth:
      secretRef:
        name: foovalue
EOF
```

> * Creates a ApisixRoute, and enable plugin `key-auth`:

Reviewer: ```suggestion
* Creates an ApisixRoute, and enables plugin `key-auth`:

```

```shell
kubectl apply -f - <<EOF
apiVersion: apisix.apache.org/v2beta3
kind: ApisixRoute
metadata:
  name: httpserver-route
spec:
  http:
  - name: rule1
    match:
      hosts:
      - httpbin.org
      paths:
      - /*
    backends:
    - serviceName: httpbin
      servicePort: 80
    authentication:
      enable: true
      type: keyAuth
EOF
```

* Requests from foo:

```shell
kubectl  exec -it -n ${namespace of Apache APISIX} ${pod of Apache APISIX}  -- curl http://127.0.0.1:9080/anything -H 'Host: httpbin.org' -H 'apikey:foo-key' -i
```

```shell
HTTP/1.1 200 OK
...
```

</details>

#### Enable `JWT Auth`

> * Creates a ApisixConsumer, and set the attributes of plugin `jwt-auth`:

Reviewer: ```suggestion
* Creates an ApisixConsumer, and set the attributes of plugin `jwt-auth`:

```

<details>
  <summary>ApisixRoute with keyAuth consumer using secret example</summary>

> * Creates an `Secret` object:

Reviewer 5: ```suggestion
* Creates a `Secret` object:

```

**Example usage**

> * Requests from jack1:

Reviewer: Just to make sure: whether the `*` signal is added by your IDE?

Author: Not, I added it.
